src/anomalib/models/dsr/torch_model.py

```diff
@@ -0,0 +1,1196 @@
+        num_embeddings: int = 4096,
+        num_hiddens: int = 128,
+        num_residual_layers: int = 2,
+        num_residual_hiddens: int = 64,
+    ):
+        super().__init__()
+
+        self.image_dim: int = 3
+        self.anomaly_map_dim: int = 2
+        self.anom_par: float = anom_par
+
+        self.discrete_latent_model = DiscreteLatentModel(
+            num_hiddens=num_hiddens,
+            num_residual_layers=num_residual_layers,
+            num_residual_hiddens=num_residual_hiddens,
+            num_embeddings=num_embeddings,
+            embedding_dim=embedding_dim,
+        )
+
+        self.image_reconstruction_network = ImageReconstructionNetwork(
+            in_channels=embedding_dim * 2,
+            num_hiddens=num_hiddens,
+            num_residual_layers=num_residual_layers,
+            num_residual_hiddens=num_residual_hiddens,
+        )
+
+        self.subspace_restriction_module_lo = SubspaceRestrictionModule(base_width=embedding_dim)
+        self.subspace_restriction_module_hi = SubspaceRestrictionModule(base_width=embedding_dim)
+
+        self.anomaly_detection_module = AnomalyDetectionModule(
+            in_channels=2 * self.image_dim, out_channels=self.anomaly_map_dim, base_width=64
+        )
+
+        self.upsampling_module = UpsamplingModule(
+            in_channels=(2 * self.image_dim) + self.anomaly_map_dim,
+            out_channels=self.anomaly_map_dim,
+            base_width=64,
+        )
+
+        for parameters in self.discrete_latent_model.parameters():
+            parameters.requires_grad = False
+
+    def load_pretrained_discrete_model_weights(self, ckpt: str) -> None:
+        self.discrete_latent_model.load_state_dict(torch.load(ckpt))
+
+    def forward(
+        self, batch: Tensor, anomaly_map_to_generate: Tensor | None = None
+    ) -> Tensor | tuple[Tensor, Tensor] | tuple[Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor]:
+        """Compute the anomaly mask from an input image.
+
+        Args:
+            batch (Tensor): Batch of input images.
+            anomaly_map_to_generate (Tensor | None): anomaly map to use to generate quantized defects.
+            If not training phase 2, should be None.
+
+        Returns:
+            If training phase 3, tensor of reconstructed anomaly map
+            If testing, tuple of:
+                - Upsampled anomaly map
+                - Image score
+            If training phase 2, tuple of:
+                - Reconstructed non-quantized hi features of defect (F~_hi)
+                - Reconstructed non-quantized lo features of defect (F~_lo)
+                - Quantized features of non defective img (Q_hi)
+                - Quantized features of non defective img (Q_lo)
+                - Object-specific-decoded image (I_spc)
+                - Computed segmentation mask (M)
+                - Resized ground-truth anomaly map (M_gt)
+        """
+        # top == lo
+
+        outputs = None
```

Comment:
It would be more readable if `outputs` is annotated. Something like 
```suggestion
        outputs: dict[str, Tensor]
```